File: RujiaLiu/chapter 4/uva213/main.py
"""
TNM AEIOU
001|0,1, 011|000111, 010|00;10;01;11, 011|001;111, 000

T, A, N, BLK, M, E
T->0,  0
N->00, 0
M->01, 1
BLK->10 2
A->000 0
E->001 1
I->010 2
O->011 3
U->100 4

build 2D matrix 
{len: index }
1: 0
2: 0 1 2 
3: 0 1 2 3 4 5 6 7 
msg fill the 2D matrix

"""

local_run = False
import sys

msg = ""
decoding = ""
code  = [[' ' for _ in range(1<<8)] for _ in range(8)]

def fill_code():
    global code, msg
    idx = 0
    for l in range(1,8):
        for i in range(1<<l)-1:
            if idx<len(msg):
                code[l][i]=msg[idx]
                idx+=1
            else:
                return
def

def main():
    global msg, decoding, code
    if local_run:
        sys.stdin = open('input.txt', 'r')
        sys.stdout = open('output.txt', 'w')
    flag_read_msg = True
    for line in sys.stdin:
        if flag_read_msg:
            msg = line.strip()
            # fill msg into 2D matrix
            fill_code()
            flag_read_msg = False
        else:
            decoding+=line.strip()
            if len(decoding)>=3 and decoding[-1]=='0' and decoding[-2]=='0' and decoding[-3]=='0':
                ret = ''
                readlen
                
                
            
if __name__ == "__main__":
    main()

File: RujiaLiu/chapter 4/uva213/main.cpp
# define LOCAL
# include<stdio.h>
# include<string.h>

int readchar(){
    for(;;){
        int ch=getchar();
        if(ch!='\n' && ch!='\r') return ch;
    }
}

int readint(int c){
    int v = 0;
    while(c--) v=v*2+readchar()-'0';
    return v;
}

int code[8][1<<8];
int readcodes(){
    memset(code, 0, sizeof(code));
    code[1][0] = readchar();
    for (int len=2; len<=7; len++){
        for (int i=0; i<(1<<len)-1; i++){
            int ch=getchar();
            if(ch==EOF) return 0;
            if(ch=='\n' || ch=='\r') return 1;
            code[len][i]=ch;
        }
    }
    return 1;
}

int main(){
    #ifdef LOCAL
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
    #endif
    while(readcodes()){
        for (;;){
            int len=readint(3);
            if (len==0) break;
            for (;;){
                int v = readint(len);
                if (v== (1<<len)-1) break;
                putchar(code[len][v]);
            }
        }
        putchar('\n');
    }
    return 0;
}